pkg/migrate/azure/dns.go

```diff
@@ -0,0 +1,1335 @@
+package azure
+
+import (
+	"bytes"
+	"context"
+	"encoding/gob"
+	"errors"
+	"fmt"
+	"reflect"
+	"strings"
+	"time"
+
+	"github.com/Azure/go-autorest/autorest/to"
+	"github.com/davecgh/go-spew/spew"
+
+	"github.com/sirupsen/logrus"
+
+	aznetwork "github.com/Azure/azure-sdk-for-go/services/network/mgmt/2018-12-01/network"
+	azdns "github.com/Azure/azure-sdk-for-go/services/preview/dns/mgmt/2018-03-01-preview/dns"
+	azprivatedns "github.com/Azure/azure-sdk-for-go/services/privatedns/mgmt/2018-09-01/privatedns"
+	azconfig "github.com/openshift/installer/pkg/asset/installconfig/azure"
+)
+
+type legacyDNSZoneSerialized struct {
+	zone       []byte
+	recordsets [][]byte
+}
+
+type legacyDNSZoneInfo struct {
+	zone       *azdns.Zone
+	recordsets []*azdns.RecordSet
+}
+
+type legacyDNSClient struct {
+	resourceGroup    string
+	public           bool
+	zonesClient      azdns.ZonesClient
+	recordsetsClient azdns.RecordSetsClient
+}
+
+func newLegacyDNSClient(session *azconfig.Session, resourceGroup string, public bool) *legacyDNSClient {
+	zonesClient := azdns.NewZonesClient(session.Credentials.SubscriptionID)
+	zonesClient.Authorizer = session.Authorizer
+
+	recordsetsClient := azdns.NewRecordSetsClient(session.Credentials.SubscriptionID)
+	recordsetsClient.Authorizer = session.Authorizer
+
+	return &legacyDNSClient{resourceGroup, public, zonesClient, recordsetsClient}
+}
+
+// safe pointer handling for int32
+func safeInt32(n *int32) int32 {
+	if n == nil {
+		return 0
+	}
+	return *n
+}
+
+// safe pointer handling for int64
+func safeInt64(n *int64) int64 {
+	if n == nil {
+		return 0
+	}
+	return *n
+}
+
+// safe pointer handling for string
+func safeString(s *string) string {
+	if s == nil {
+		return ""
+	}
+	return *s
+}
+
+// safe pointer handling for array of string
+func safeStringArray(s *[]string) string {
+	if s == nil {
+		return ""
+	}
+	return (*s)[0]
+}
+
+// Takes a subscription ID and parses the resource group out of it
+func idToResourceGroup(id string) string {
+	parts1 := strings.Split(id, "resourceGroups/")
+	if len(parts1) < 2 {
+		return ""
+	}
+
+	parts2 := strings.Split(parts1[1], "/providers")
+	if len(parts2) < 2 {
+		return ""
+	}
+
+	return parts2[0]
+}
```

Comment:
can you add a unit test for this..?
also this needs a little bit more documentation on the function.

Reply:
I've updated this function a bit, and also commented it a bit more.